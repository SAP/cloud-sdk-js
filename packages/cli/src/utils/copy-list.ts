/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

export interface CopyDescriptor {
  sourcePath: string | URL;
  fileName: string;
}
