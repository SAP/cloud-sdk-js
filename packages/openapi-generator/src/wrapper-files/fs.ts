import { promises, existsSync } from 'fs';
import { join } from 'path';
import { codeBlock } from '@sap-cloud-sdk/util';
const { writeFile } = promises;
const fsCopyFile = promises.copyFile;

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Write a file generated by the SAP Cloud SDK for JavaScript.
 * @param directoryPath Path of the directory to write to.
 * @param fileName Name of the file to write
 * @param content Content to be written to the file. A copyright statement will be added to this.
 * @param overwrite Whether or not existing files should be overwritten.
 * @param withCopyright Whether the generated file contains the copyright information.
 */
export async function createFile(
  directoryPath: string,
  fileName: string,
  content: string,
  overwrite: boolean,
  withCopyright = true
): Promise<void> {
  return writeFile(
    join(directoryPath, fileName),
    withCopyright ? wrapContent(content) : content,
    {
      encoding: 'utf8',
      flag: overwrite ? 'w' : 'wx'
    }
  );
}

function wrapContent(content: string): string {
  return (
    codeBlock`
/*
 * Copyright (c) ${new Date().getFullYear()} SAP SE or an SAP affiliate company. All rights reserved.
 *
 * This is a generated file powered by the SAP Cloud SDK for JavaScript.
 */
${content}
` + '\n'
  );
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Copy a file from a given path.
 * @param src Path to the source file.
 * @param dest Path to the destination file
 * @param overwrite Whether or not existing files should be overwritten.
 */
export async function copyFile(
  src: string,
  dest: string,
  overwrite: boolean
): Promise<void> {
  if (!overwrite && (await existsSync(dest))) {
    return;
  }
  return fsCopyFile(src, dest);
}
