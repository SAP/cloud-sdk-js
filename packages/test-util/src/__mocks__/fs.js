let mockedFS = jest.genMockFromModule('fs');
const originalFS = jest.requireActual('fs');

let mockingIsOn = false;
let mockReadDirSync = [];
let mockReadFileSync = '';

function switchMockOn() {
  mockingIsOn = true;
}

function switchMockOff() {
  mockingIsOn = false;
}

function setReadDirSync(files) {
  mockReadDirSync = files;
}

function setReadFileSync(filePath, file) {
  mockReadFileSync = file;
}

function readdirSync(directoryPath) {
  if (!mockingIsOn) {
    return originalFS.readdirSync(directoryPath);
  }
  return mockReadDirSync;
}

function readFileSync(filePath, encoding) {
  if (!mockingIsOn) {
    return originalFS.readFileSync(filePath, encoding);
  }
  return mockReadFileSync;
}

//Take default for all which is not mocked
for (var prop in mockedFS) {
  if (mockedFS.hasOwnProperty(prop) && originalFS.hasOwnProperty(prop)) {
    if (typeof mockedFS[prop] === 'function') {
      mockedFS[prop] = originalFS[prop];
    }
  }
}

//set mock functions
mockedFS.switchMockOn = switchMockOn;
mockedFS.switchMockOff = switchMockOff;

mockedFS.setReadDirSync = setReadDirSync;
mockedFS.readdirSync = readdirSync;

mockedFS.setReadFileSync = setReadFileSync;
mockedFS.readFileSync = readFileSync;

module.exports = mockedFS;
