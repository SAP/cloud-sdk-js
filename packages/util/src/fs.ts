import { existsSync, PathLike, readdirSync, readFileSync } from 'fs';
import { resolve } from 'path';
import { createLogger } from './logger';

const logger = createLogger({
  package: 'util',
  messageContext: 'fs'
});

/**
 * @internal
 */
export function findProjectRoot(path: string, lastPath: string = path): string {
  if (!path) {
    return lastPath;
  }

  const inProject =
    readdirSync(path).includes('package.json') ||
    readdirSync(path).includes('node_modules') ||
    path.includes('node_modules');
  if (!inProject) {
    return lastPath;
  }

  return findProjectRoot(resolve(path, '..'), path);
}

/**
 * Read a JSON file from the file system.
 * @param path - The path to the JSON file.
 * @returns An object parsed from the JSON file.
 */
export function readJSON(path: PathLike): { [key: string]: any } {
  if (existsSync(path)) {
    return JSON.parse(readFileSync(path, 'utf8'));
  }
  logger.warn(`File "${path}" does not exist, return empty object.`);
  return {};
}
