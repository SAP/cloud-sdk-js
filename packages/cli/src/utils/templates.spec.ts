/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
import { resolve, relative, sep } from 'path';
import fs from 'fs-extra';
import {
  deleteAsync,
  getCleanProjectDir,
  getTestOutputDir,
  TimeThresholds
} from '../../test/test-utils';
import {
  copyFiles,
  findConflicts,
  getCopyDescriptors,
  getTemplatePaths
} from './templates';
import { CopyDescriptor } from './copy-list';

const testOutputDir = getTestOutputDir(__filename);

describe('Templates Utils', () => {
  beforeAll(async () => {
    deleteAsync(testOutputDir, 3);
  }, TimeThresholds.EXTRA_SHORT);

  it(
    'should return information which files to copy where',
    () => {
      const initCopyInfo = getCopyDescriptors(
        'targetDir',
        getTemplatePaths(['init'])
      );
      expect(
        initCopyInfo.map(copyInfo => copyInfoToPathArray(copyInfo)).sort()
      ).toMatchSnapshot();

      const appRouterCopyInfo = getCopyDescriptors(
        'targetDir',
        getTemplatePaths(['add-approuter'])
      );
      expect(
        appRouterCopyInfo.map(info => copyInfoToPathArray(info).sort())
      ).toMatchSnapshot();
    },
    TimeThresholds.EXTRA_SHORT
  );

  it(
    'should find conflicts',
    async () => {
      const projectDir = await getCleanProjectDir(
        testOutputDir,
        'find-conflicts'
      );
      await fs.writeFile(resolve(projectDir, 'manifest.yml'), 'foobar');
      findConflicts(
        getCopyDescriptors(projectDir, getTemplatePaths(['init'])),
        true
      );
      try {
        await fs.stat(resolve(projectDir, 'manifest.yml'));
      } catch (e) {
        expect(e.message).toMatch(/no such file or directory.*manifest.yml/);
      }
    },
    TimeThresholds.EXTRA_SHORT
  );

  it(
    'should copy files locally',
    async () => {
      const projectDir = await getCleanProjectDir(
        testOutputDir,
        'copy-files-locally'
      );
      await copyFiles(
        getCopyDescriptors(projectDir, getTemplatePaths(['init'])),
        {}
      );
      return fs
        .readdir(projectDir)
        .then(value => expect(value).toMatchSnapshot);
    },
    TimeThresholds.EXTRA_SHORT
  );

  // TODO:
  // it('should copy files remotely', async () => {
  //   const projectDir = getCleanProjectDir(testOutputDir, 'copy-files-remotely');
  // });

  function copyInfoToPathArray(copyInfo: CopyDescriptor): string[] {
    const filePathBeginnginFromTargetDir = relative(
      resolve('targetDir'),
      copyInfo.fileName
    );
    const filePathAsList = filePathBeginnginFromTargetDir.split(sep);
    return filePathAsList;
  }
});
