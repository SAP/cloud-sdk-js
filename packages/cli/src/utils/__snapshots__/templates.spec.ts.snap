// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Templates Utils should return information which files to copy where 1`] = `
Array [
  Array [
    ".pipeline",
    "config.yml",
  ],
  Array [
    "Jenkinsfile",
  ],
  Array [
    "credentials.json",
  ],
  Array [
    "deployment",
    ".gitkeep",
  ],
  Array [
    "manifest.yml",
  ],
  Array [
    "s4hana_pipeline",
    "reports",
    "backend-integration",
    ".gitkeep",
  ],
  Array [
    "s4hana_pipeline",
    "reports",
    "backend-unit",
    ".gitkeep",
  ],
  Array [
    "s4hana_pipeline",
    "reports",
    "coverage-reports",
    "backend-integration",
    ".gitkeep",
  ],
  Array [
    "s4hana_pipeline",
    "reports",
    "coverage-reports",
    "backend-unit",
    ".gitkeep",
  ],
  Array [
    "systems.json",
  ],
]
`;

exports[`Templates Utils should return information which files to copy where 2`] = `
Array [
  Array [
    "approuter",
    "manifest.yml",
  ],
  Array [
    "approuter",
    "package.json",
  ],
  Array [
    "approuter",
    "xs-app.json",
  ],
  Array [
    "approuter",
    "xs-security.json",
  ],
]
`;
